starlite/datastructures/provide.py

```diff
@@ -1,27 +1,79 @@
+from inspect import Parameter, Signature
 from typing import TYPE_CHECKING, Any, Optional, cast
 
+from anyio import Lock
+
 from starlite.types import Empty
 from starlite.utils.sync import AsyncCallable, is_async_callable
 
 if TYPE_CHECKING:
     from typing import Type
 
+    from starlite.connection import Request
     from starlite.signature import SignatureModel
     from starlite.types import AnyCallable
 
 
+def _wrap_cache_for_request(fn: "AnyCallable", override_name: Optional[str]) -> "AnyCallable":
```

Comment:
this is a good place to use a `Generic` parameter T, and later to use `ParamSpec` P. See for an example: `starlite/utils/sync.py`.